clap-utils/src/keypair.rs

```diff
@@ -196,10 +198,13 @@ impl DefaultSigner {
     /// `bulk_signers` is a vector of signers, all of which are optional. If any
     /// of those signers is `None`, then the default signer will be loaded; if
     /// all of those signers are `Some`, then the default signer will not be
-    /// loaded.
+    /// loaded. If multiple equivalent (same pub key) signers are provided - only
+    /// one of those will be returned in the result, such that NullSigner(s)
+    /// always get lower priority.
     ///
     /// The returned value includes all of the `bulk_signers` that were not
-    /// `None`, and maybe the default signer, if it was loaded.
+    /// `None`, and maybe the default signer (if it was loaded). There is no
+    /// guarantees on resulting signers ordering.
```

Comment:
This is needed to allow specifying same signer seamlessly multiple times (with some of those "duplicates" being NullSigner(s)). One case I found this to be helpful in is doing something like this:
```
solana program deploy --sign-only --fee-payer <OFFLINE_SIGNER_PUB_KEY> --program <PROGRAM_SIGNER_PUB_KEY> --upgrade-authority <OFFLINE_SIGNER> --buffer <BUFFER_PUB_KEY> --blockhash <VALUE> --max-len <VALUE> --min-rent-balance <VALUE>
```
where `<OFFLINE_SIGNER>` can have same pubkey as `<OFFLINE_SIGNER_PUB_KEY>` (and even perhaps `<PROGRAM_SIGNER_PUB_KEY>`).

Reply:
One reason (closely-related to this PR) is the use-case of specifying both `--upgrade-authority <OFFLINE_SIGNER_PUB_KEY>` and `--signer <OFFLINE_SIGNER_SIGNATURE>` when doing `solana program upgrade`:
```bash
solana program upgrade --fee-payer /Users/norwnd/.config/solana/id.json --program-id BgY6WizXe2Hfrq71TjjvV2ntjiMX3tjHKAARvg47SJUC --upgrade-authority GuKy3G22F4avoPKGtpqBEuoQYLEWaCRmncGGc5Q7JoPL --buffer C1o2XLFAU6EqZuMPVfW9C2gFEtBxMuXaNkEYLyNtsbiU --blockhash qyof8GDHGrVwUqtiDAFfit9MY62Xub2i1cEAnVXpQ62 --signer GuKy3G22F4avoPKGtpqBEuoQYLEWaCRmncGGc5Q7JoPL=Qonons7tm1oRxR2akT4cVu9HyXQjah2DxAzS8Ht7RwU3Gx1Nk2SUTv9X9LMa8mYFrPSRZbpYDsNVPG5BNNLGsh7
```
since `OFFLINE_SIGNER_PUB_KEY` will convert to `NullSiger` we need to differentiate between these 2 signers (preffering the most complete of these 2). Which is why I didn't create a separate PR for it - but can potentially move these changes into "a separate PR with preliminary changes" this PR depends on, if that helps with anything.

Alternative way to address this would be to move this logic into `("upgrade", Some(matches)) => {` possibly making it specific to only certain params (only those we want to allow this feature for), but it seems as an inferior approach to me.

Reply:
In the CLI, this is all done for you with `signer_of`. If you follow the code to `signer_from_path_with_config`, you'll see that if a pubkey is provided, the function will then check the `--signer` args to see if a signature is provided, and use that signature for any transaction.

For example, see how `process_transfer` is using the different signers: https://github.com/solana-labs/solana/blob/2c033e05e5aa43fe76a31f18a2035c25b1d26d63/cli/src/wallet.rs#L871

So the code is all in there for you already!

Reply:
Right, what you describe here solves it specifically for `--signer` argument, 

I've [described here](https://github.com/solana-labs/solana/pull/33860#discussion_r1416909842) why we might want to solve it for other parameters too, wdyt ?

Reply:
Changed to get rid of this NullSigner-based approach I've suggested above (see latest PR version).